gax/src/main/java/com/google/api/gax/retrying/ExponentialRetryAlgorithm.java

```diff
@@ -100,19 +100,31 @@ public TimedAttemptSettings createNextAttempt(TimedAttemptSettings prevSettings)
           (long) (settings.getRetryDelayMultiplier() * prevSettings.getRetryDelay().toMillis());
       newRetryDelay = Math.min(newRetryDelay, settings.getMaxRetryDelay().toMillis());
     }
+    Duration randomDelay = Duration.ofMillis(nextRandomLong(newRetryDelay));
 
     // The rpc timeout is determined as follows:
     //     attempt #0  - use the initialRpcTimeout;
-    //     attempt #1+ - use the calculated value.
+    //     attempt #1+ - use the calculated value or the time remaining in totalTimeout.
     long newRpcTimeout =
         (long) (settings.getRpcTimeoutMultiplier() * prevSettings.getRpcTimeout().toMillis());
     newRpcTimeout = Math.min(newRpcTimeout, settings.getMaxRpcTimeout().toMillis());
 
+    // The totalTimeout could be zero if a callable is only using maxAttempts to limit retries.
+    // If set, calculate time remaining in the totalTimeout since the start, taking into account the
+    // next attempt's delay, in order to truncate the RPC timeout should it exceed the totalTimeout.
+    if (!settings.getTotalTimeout().isZero()) {
+      Duration timeElapsed =
+          Duration.ofNanos(clock.nanoTime())
+              .minus(Duration.ofNanos(prevSettings.getFirstAttemptStartTimeNanos()));
+      Duration timeLeft = globalSettings.getTotalTimeout().minus(timeElapsed).minus(randomDelay);
+      newRpcTimeout = Math.min(newRpcTimeout, timeLeft.toMillis());
```

Comment:
What would happen if `newRpcTimeout <= 0`?

Reply:
If `newRpcTimeout <= 0`, the subsequent check by `shouldRetry` would prevent another attempt from being made, because they do a similar check to see if the current time + the proposed random delay would exceed the `totalTimeout`:
```java
long totalTimeSpentNanos =
        clock.nanoTime()
            - nextAttemptSettings.getFirstAttemptStartTimeNanos()
            + nextAttemptSettings.getRandomizedRetryDelay().toNanos();

    // If totalTimeout limit is defined, check that it hasn't been crossed
    if (totalTimeout > 0 && totalTimeSpentNanos > totalTimeout) {
      return false;
    }
```

We could add some logic to handle `if (timeLeft.isNegative() || timeLeft.isZero())`, but I'm not sure what we'd set it to...maybe allow `newRpcTimeout` to remain unchanged by `timeLeft`?

Reply:
After our offline conversation, I've decided to just document the likely behavior. LMK if you feel strongly otherwise.